### ISSUE TITLE
Parallelisation of `massflow`

### ISSUE THREAD

**Issue by scheidan** [santiago-sanitation-systems/Santiago.jl#11]
b37fea6 removes the incorrect multi-threading implementation of `massflow!`.

The massflow calculation are costly. A Monte Carlos simulation with 100 samples (`massflow_summary!(sys, input_masses, n=100)`) takes around 250 ms.

Possible approaches are:
- use `Distributed.@distributed` for parallelisation on different cores. Needs a lot of `@everywhere`
- correct multithreading, see [here](https://docs.julialang.org/en/v1/manual/parallel-computing/#Side-effects-and-mutable-function-arguments-1). Could have quite some overhead.
- compute massflows only on demand for the selected systems.

**Comment by scheidan** [comment]
Implementation of a `ProgressMeter` would be useful too

**Comment by scheidan** [comment]
With `ProgressMeter.jl`:
```Julia
function runall!(allSys, M_in; n=100)
    # define a separate random number generator for each thread
    mt = MersenneTwister(1)
    rngs = [mt; accumulate(Future.randjump, fill(big(10)^20, Threads.nthreads()-1), init=mt)]

    p = Progress(length(allSys), dt=1, desc="Compute massflows:")
    Threads.@threads for s in allSys
        massflow_summary!(s, M_in, n=n, rng=rngs[Threads.threadid()])
        next!(p)
    end
end
```

**Comment by scheidan** [comment]
Some timings on my notebook (probably with 4 cores):
```
------
Number of threads: 1
228 systems with 100 MC sampels
Min: TrialEstimate(16.997 s)
Median: TrialEstimate(16.997 s)
Allocs: 148987904
Memory: 15104800496
------
Number of threads: 2
228 systems with 100 MC sampels
Min: TrialEstimate(10.821 s)
Median: TrialEstimate(10.821 s)
Allocs: 148987539
Memory: 15104792816
------ 
Number of threads: 3
228 systems with 100 MC sampels
Min: TrialEstimate(8.639 s)
Median: TrialEstimate(8.639 s)
Allocs: 148987576
Memory: 15104821584
------ 
Number of threads: 4
228 systems with 100 MC sampels
Min: TrialEstimate(8.640 s)
Median: TrialEstimate(8.640 s)
Allocs: 148986790
Memory: 15104769536
------
Number of threads: 5
228 systems with 100 MC sampels
Min: TrialEstimate(8.503 s)
Median: TrialEstimate(8.503 s)
Allocs: 148987187
Memory: 15104831312
------
Number of threads: 6
228 systems with 100 MC sampels
Min: TrialEstimate(7.723 s)
Median: TrialEstimate(7.723 s)
Allocs: 148987213
Memory: 15104855968
------
Number of threads: 7
228 systems with 100 MC sampels
Min: TrialEstimate(9.030 s)
Median: TrialEstimate(9.030 s)
Allocs: 148987263
Memory: 15104883616
------
Number of threads: 8
228 systems with 100 MC sampels
Min: TrialEstimate(8.235 s)
Median: TrialEstimate(8.235 s)
Allocs: 148987592
Memory: 15104938368
```

**Comment by scheidan** [comment]
Some more timings for build systems:
```
------
1
 
Number of threads: 1
Min: TrialEstimate(199.007 s)
Median: TrialEstimate(199.007 s)
Allocs: 1096256293
Memory: 124403643360
------
2
 
Number of threads: 2
Min: TrialEstimate(216.600 s)
Median: TrialEstimate(216.600 s)
Allocs: 1123679650
Memory: 125368013984
------
3
 
Number of threads: 3
Min: TrialEstimate(165.315 s)
Median: TrialEstimate(165.315 s)
Allocs: 1111649365
Memory: 124797470304
------
4
 
Number of threads: 4
Min: TrialEstimate(215.034 s)
Median: TrialEstimate(215.034 s)
Allocs: 1119399695
Memory: 125123287648
```

**Comment by scheidan** [comment]
Implemented by e24af76
